### PR TITLE
Python: Support for adding columns

### PR DESCRIPTION
**Pull request by hililiwei**
Under the umbrella of #7875

### What is the purpose of the change

Added the ability to add columns to PyIceberg.

### CODE REVIEW

File: python/pyiceberg/schema.py [hunk 1]
```diff
@@ -1367,3 +1382,258 @@ def _(file_type: DecimalType, read_type: IcebergType) -> IcebergType:
             raise ResolveError(f"Cannot reduce precision from {file_type} to {read_type}")
     else:
         raise ResolveError(f"Cannot promote an decimal to {read_type}")
+
+
+class UpdateSchema(ABC):
```
Comment (Fokko) [inline python/pyiceberg/schema.py:1387]: Do we need this ABC? Feels a bit Java'ish, also the naming is confusing `UpdateSchema` and `SchemaUpdate` as the implementation

File: python/pyiceberg/table/__init__.py [hunk 2]
```diff
@@ -152,6 +154,27 @@ def set_properties(self, **updates: str) -> Transaction:
         """
         return self._append_updates(SetPropertiesUpdate(updates=updates))
 
+    def add_column(
```
Comment (Fokko) [inline python/pyiceberg/table/__init__.py:157]: I don't think we want to have this API on the Transaction itself. I like what @rdblue is suggesting [here](https://github.com/apache/iceberg/pull/6323#discussion_r1217020366).

This would look like:
```python
# When you want to do multiple things
with table.transaction() as t:
    t.updateSchema().addColumn("x", IntegerType()).commit()
    t.set_property(updated_at=str(datetime.now())

# or when you don't want to use a transaction:
t.updateSchema().addColumn("x", IntegerType()).commit()    
```

The `.commit()` will accumulate all the changes to the schema, and combine this into a single `AddSchemaUpdate` that's being added to the `self._updates`. This way we don't need `self._schema_update` which seems to be a little brittle since we set it to null again.

File: python/tests/conftest.py [hunk 3]
```diff
@@ -153,13 +153,19 @@ def table_schema_nested() -> Schema:
         NestedField(
             field_id=11,
             name="location",
-            field_type=ListType(
-                element_id=12,
-                element_type=StructType(
+            field_type=MapType(
```
Comment (Fokko) [inline python/tests/conftest.py:156]: Maybe better to not change this fixture, and create a new schema?

File: python/pyiceberg/schema.py (hunk 1 above)
Comment (hililiwei) [reply python/pyiceberg/schema.py]: I moved `SchemaUpdate`  to `table/__init__.py` as an internal private class, and do you think that's better?

The current names reference the java code names, so what do we change them to?

Comment (hililiwei) [reply python/pyiceberg/schema.py]: > Do we need this ABC? Feels a bit Java'ish

I dropped that `ABC`. Does keeping `ABC` make it clearer that it is an interface?

File: python/pyiceberg/table/__init__.py (hunk 2 above)
Comment (hililiwei) [reply python/pyiceberg/table/__init__.py]: I have made changes and currently support both of the above usage.

But do we need to support the following way?

```
txn = table.transaction()
txn.update_schema().add_column(name="x", type_var=IntegerType(), doc="doc")
txn.commit_transaction()
````

File: python/tests/conftest.py (hunk 3 above)
Comment (hililiwei) [reply python/tests/conftest.py:207]: reverted.

File: python/pyiceberg/schema.py [hunk 4]
```diff
@@ -1082,44 +1101,40 @@ def build_position_accessors(schema_or_type: Union[Schema, IcebergType]) -> Dict
...
 
-    def __init__(self, start: int = 1) -> None:
+    def __init__(self, start: int = 1, next_id_func: Optional[Callable[[], int]] = None) -> None:
         self.counter = itertools.count(start)
         self.reserved_ids = {}
+        self.next_id_func = next_id_func
 
     def _get_and_increment(self) -> int:
+        if self.next_id_func:
+            return self.next_id_func()
```
[the remark this answers is not in the document]
Comment (hililiwei) [reply python/pyiceberg/schema.py]: done.

File: python/pyiceberg/schema.py [hunk 5]
```diff
@@ -1082,44 +1101,40 @@ def build_position_accessors(schema_or_type: Union[Schema, IcebergType]) -> Dict
...
-
-    def struct(self, struct: StructType, field_results: List[int]) -> int:
-        return max(field_results)
-
-    def field(self, field: NestedField, field_result: int) -> int:
-        return max(field.field_id, field_result)
-
-    def list(self, list_type: ListType, element_result: int) -> int:
-        return element_result
+@singledispatch
```
[the remark this answers is not in the document]
Comment (hililiwei) [reply python/pyiceberg/schema.py]: done.

File: python/tests/test_schema.py [hunk 6]
```diff
@@ -247,78 +247,6 @@ def test_schema_find_field_type_by_id(table_schema_simple: Schema) -> None:
     assert index[3] == NestedField(field_id=3, name="baz", field_type=BooleanType(), required=False)
 
 
-def test_index_by_id_schema_visitor(table_schema_nested: Schema) -> None:
```
Comment (hililiwei) [inline python/tests/test_schema.py:250]: Repeat with `test_schema_index_by_id_visitor`

File: python/pyiceberg/table/__init__.py (hunk 2 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py]: Thinking of it, I think the transaction needs to be on a catalog level.

File: python/pyiceberg/table/__init__.py [hunk 7]
```diff
@@ -841,3 +868,210 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        self._allow_incompatible_changes: bool = False
+        self._case_sensitive: bool = True
+
+    def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        self._case_sensitive = case_sensitive
+        return self
+
+    def add_column(
+        self, name: str, type_var: IcebergType, doc: Optional[str] = None, parent: Optional[str] = None
+    ) -> UpdateSchema:
```
Comment (Fokko) [inline python/pyiceberg/table/__init__.py:895]: What do you think of keeping this in a single function?
```suggestion
    def add_column(
        self, name: str, type_var: IcebergType, doc: Optional[str] = None, parent: Optional[str] = None, required: bool = False
    ) -> UpdateSchema:
```
The majority of the column being added will be optional.

File: python/mkdocs/docs/api.md [hunk 8]
```diff
@@ -146,6 +146,29 @@ catalog.create_table(
...
 
+### Update table schema
+
+Add new columns through the `Transaction` or `UpdateSchema` API:
+
+Use the Transaction API:
+
+```python
+with table.transaction() as transaction:
+    transaction.update_schema().add_column("x", IntegerType(), "doc").commit()
```
Comment (Fokko) [inline python/mkdocs/docs/api.md:157]: The context manager will do the commit for you:
```suggestion
    transaction.update_schema().add_column("x", IntegerType(), "doc")
```

File: python/pyiceberg/table/__init__.py [hunk 9]
```diff
@@ -841,3 +868,210 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        return self
+
+    def add_required_column(
+        self, name: str, type_var: IcebergType, doc: Optional[str] = None, parent: Optional[str] = None
+    ) -> UpdateSchema:
+        if "." in name:
+            raise ValueError(f"Cannot add column with ambiguous name: {name}")
+
+        if not self._allow_incompatible_changes:
+            raise ValueError(f"Incompatible change: cannot add required column: {name}")
```
Comment (Fokko) [inline python/pyiceberg/table/__init__.py:909]: Can you add a comment here that this will be possible with spec V3, which sets an initial value?

File: python/pyiceberg/table/__init__.py [hunk 10]
```diff
@@ -841,3 +868,210 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        self._internal_add_column(parent, name, False, type_var, doc)
+        return self
+
+    def allow_incompatible_changes(self) -> UpdateSchema:
+        self._allow_incompatible_changes = True
+        return self
+
+    def apply(self) -> Schema:
```
Comment (Fokko) [inline python/pyiceberg/table/__init__.py:918]: Should we make this one private?
```suggestion
    def _apply(self) -> Schema:
```

File: python/pyiceberg/schema.py [hunk 11]
```diff
@@ -1082,45 +1100,23 @@ def build_position_accessors(schema_or_type: Union[Schema, IcebergType]) -> Dict
...
 
 class _SetFreshIDs(PreOrderSchemaVisitor[IcebergType]):
     """Traverses the schema and assigns monotonically increasing ids."""
 
-    counter: itertools.count  # type: ignore
     reserved_ids: Dict[int, int]
 
-    def __init__(self, start: int = 1) -> None:
+    def __init__(self, start: int = 1, next_id_func: Optional[Callable[[], int]] = None) -> None:
         self.counter = itertools.count(start)
```
Comment (Fokko) [inline python/pyiceberg/schema.py:1114]: This one can go now

File: python/pyiceberg/table/__init__.py [hunk 12]
```diff
@@ -841,3 +868,210 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        # update tracking for moves
+        self._added_name_to_id[full_name] = new_id
+
+        new_type = assign_fresh_schema_ids(type_var, self.assign_new_column_id)
+        field = NestedField(new_id, name, new_type, not is_optional, doc)
+
+        self._adds.setdefault(parent_id, []).append(field)
+
+    def assign_new_column_id(self) -> int:
```
Comment (Fokko) [inline python/pyiceberg/table/__init__.py:983]: What do you think of using a counter here as well?

File: python/pyiceberg/table/__init__.py [hunk 13]
```diff
@@ -841,3 +868,210 @@ def to_ray(self) -> ray.data.dataset.Dataset:
         import ray
 
         return ray.data.from_arrow(self.to_arrow())
+
+
+class _SchemaUpdate(UpdateSchema):
+    def __init__(self, schema: Schema, table: Optional[Table] = None, last_column_id: Optional[int] = None):
```
Comment (Fokko) [inline python/pyiceberg/table/__init__.py:874]: Looking at the code, the `table` is always passed in, so maybe better to make that one required
```suggestion
    def __init__(self, schema: Schema, table: Table, last_column_id: Optional[int] = None):
```

Comment (hililiwei) [reply python/pyiceberg/table/__init__.py:874]: Currently in the test, no `table` are passed in

File: python/pyiceberg/table/__init__.py (hunk 9 above)
Comment (hililiwei) [reply python/pyiceberg/table/__init__.py:909]: I think so. I added a comment stating that in versions 1 and 2, adding required columns is not allowed by default because there is no initial value.

File: python/pyiceberg/schema.py (hunk 1 above)
Comment (hililiwei) [reply python/pyiceberg/schema.py]: removed.

File: python/pyiceberg/schema.py [hunk 14]
```diff
@@ -1082,45 +1100,23 @@ def build_position_accessors(schema_or_type: Union[Schema, IcebergType]) -> Dict
...
         self.reserved_ids = {}
+        self.next_id_func = next_id_func if next_id_func is not None else lambda: next(self.counter)
```
[the remark this answers is not in the document]
Comment (hililiwei) [reply python/pyiceberg/schema.py]: It can't be used this way. `_get_and_increment()` always going to be 1.

File: python/pyiceberg/table/__init__.py [hunk 15]
```diff
@@ -841,3 +868,239 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+            self._last_column_id = itertools.count(schema.highest_field_id + 1)
+
+        self._identifier_field_names = schema.column_names
+        self._adds: Dict[int, List[NestedField]] = {}
+        self._added_name_to_id: Dict[str, int] = {}
+        self._id_to_parent: Dict[int, str] = {}
+        self._allow_incompatible_changes: bool = False
+        self._case_sensitive: bool = True
+
+    def __exit__(self, _: Any, value: Any, traceback: Any) -> None:
```
[the remark this answers is not in the document]
Comment (hililiwei) [reply python/pyiceberg/table/__init__.py:901]: One question:

For:
```
with table.transaction() as transaction:
   transaction.update_schema().add_column("x1", IntegerType(), "doc").add_column("x2", IntegerType(), "doc").commit()
# now the commit happens because the transaction closes
```

There is no need to return the table, and there is no way to return the refreshed table.

[the remark this answers is not in the document]
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:901]: The reference to the table is refreshed:
```python
        table_update_response = self._table.catalog._commit_table(  # pylint: disable=W0212
            CommitTableRequest(
                identifier=self._table.identifier[1:],
                updates=[
                    AddSchemaUpdate(schema=new_schema, last_column_id=new_schema.highest_field_id),
                    SetCurrentSchemaUpdate(schema_id=-1),
                ],
            )
        )

        self._table.metadata = table_update_response.metadata
        self._table.metadata_location = table_update_response.metadata_location
```

So strictly it is not needed, and the table is already updated. I noticed this when checking the integration tests. With the context manager this is indeed not possible, I'm also fine with leaving this out.

**Review by Fokko — APPROVED** [review]

**Comment by Fokko** [comment]
Thanks @hililiwei for working on this 👍🏻

File: python/mkdocs/docs/api.md [hunk 16]
```diff
@@ -146,6 +146,38 @@ catalog.create_table(
 )
 ```
 
+### Update table schema
+
+Add new columns through the `Transaction` or `UpdateSchema` API:
+
+Use the Transaction API:
```
Comment (rdblue) [inline python/mkdocs/docs/api.md:153]: @Fokko, I don't think that we should encourage people to use the transaction API for single operations like this. It's much simpler to avoid using the transaction right?

File: python/pyiceberg/schema.py [hunk 17]
```diff
@@ -201,7 +203,7 @@ def find_type(self, name_or_id: Union[str, int], case_sensitive: bool = True) ->
 
     @property
     def highest_field_id(self) -> int:
-        return visit(self.as_struct(), _FindLastFieldId())
+        return max(self._lazy_id_to_name.keys(), default=0)
```
Comment (rdblue) [inline python/pyiceberg/schema.py:206]: This seems unrelated, though fine.

File: python/pyiceberg/schema.py [hunk 18]
```diff
@@ -15,6 +15,8 @@
 # specific language governing permissions and limitations
 # under the License.
 # pylint: disable=W0511
+from __future__ import annotations
```
Comment (rdblue) [inline python/pyiceberg/schema.py:18]: @Fokko, what is our policy on using this vs using strings? How do we choose?

File: python/pyiceberg/schema.py [hunk 19]
```diff
@@ -890,6 +892,22 @@ def __init__(self) -> None:
         self._field_names: List[str] = []
         self._short_field_names: List[str] = []
 
+    def before_map_key(self, key: NestedField) -> None:
+        self.before_field(key)
```
Comment (rdblue) [inline python/pyiceberg/schema.py:896]: Isn't this the default implementation?

File: python/pyiceberg/table/__init__.py [hunk 20]
```diff
@@ -128,6 +144,25 @@ def _append_updates(self, *new_updates: TableUpdate) -> Transaction:
...
+
+        Raises:
+            ValueError: When the type of requirement is not unique.
+
+        Returns:
+            Transaction object with the new requirements appended.
+        """
+        for requirement in new_requirements:
+            type_new_requirement = type(requirement)
+            if any(type(update) == type_new_requirement for update in self._updates):
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:161]: Should this be `_updates` or `_requirements`?

File: python/pyiceberg/table/__init__.py [hunk 21]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+            case_sensitive: When false case is not considered in column name comparisons.
+
+        Returns:
+            This for method chaining
+        """
+        self._case_sensitive = case_sensitive
+        return self
+
+    def add_column(
+        self, name: str, type_var: IcebergType, doc: Optional[str] = None, parent: Optional[str] = None, required: bool = False
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:937]: Should we support `Tuple[str]` for name? That way we don't need `parent`.

File: python/pyiceberg/table/__init__.py [hunk 22]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+            raise ValueError(f"Cannot add column with ambiguous name: {name}")
+
+        if required and not self._allow_incompatible_changes:
+            # Table format version 1 and 2 cannot add required column because there is no initial value
+            raise ValueError(f"Incompatible change: cannot add required column: {name}")
+
+        self._internal_add_column(parent, name, not required, type_var, doc)
+        return self
+
+    def allow_incompatible_changes(self) -> UpdateSchema:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:961]: What about adding these boolean options to the `update_schema(...)` method?

```python
table.update_schema(allow_incompatible_changes=True, case_sensitive=False).add_column(...).commit()
```

File: python/pyiceberg/table/__init__.py [hunk 23]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        if self._transaction is not None:
+            self._transaction._append_updates(*updates)  # pylint: disable=W0212
+            self._transaction._append_requirements(*requirements)  # pylint: disable=W0212
+        else:
+            table_update_response = self._table.catalog._commit_table(  # pylint: disable=W0212
+                CommitTableRequest(identifier=self._table.identifier[1:], updates=updates, requirements=requirements)
+            )
+            self._table.metadata = table_update_response.metadata
+            self._table.metadata_location = table_update_response.metadata_location
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:987]: I don't think these should be managed externally. Can `_table` keep track of these and run the commit instead?

File: python/pyiceberg/schema.py (hunk 17 above)
Comment (Fokko) [reply python/pyiceberg/schema.py:206]: Should have been a separate PR indeed, but it is a very welcome change

File: python/pyiceberg/schema.py (hunk 18 above)
Comment (Fokko) [reply python/pyiceberg/schema.py:18]: I can add that to [the `Type annotations` section](https://py.iceberg.apache.org/contributing/#type-annotations). I prefer the `from __future__ import annotations` since it introduces less visual noise.

File: python/pyiceberg/schema.py (hunk 19 above)
Comment (Fokko) [reply python/pyiceberg/schema.py:896]: Yes, looks like it. Let me clean that up

File: python/pyiceberg/schema.py [hunk 24]
```diff
@@ -1082,45 +1100,23 @@ def build_position_accessors(schema_or_type: Union[Schema, IcebergType]) -> Dict
...
     """Traverses the schema and assigns monotonically increasing ids."""
 
-    counter: itertools.count  # type: ignore
     reserved_ids: Dict[int, int]
 
-    def __init__(self, start: int = 1) -> None:
-        self.counter = itertools.count(start)
+    def __init__(self, next_id_func: Optional[Callable[[], int]] = None) -> None:
         self.reserved_ids = {}
+        counter = itertools.count(1)
```
[the remark this answers is not in the document]
Comment (Fokko) [reply python/pyiceberg/schema.py:1115]: The `start` argument wasn't actually exposed:
```python
def assign_fresh_schema_ids(schema: Schema) -> Schema:
    """Traverses the schema, and sets new IDs."""
    return pre_order_visit(schema, _SetFreshIDs())
```

The `_SetFreshIDs()` is private to the module, so we're safe here.

File: python/pyiceberg/table/__init__.py (hunk 20 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:161]: auch, copy-paste. Thanks!

File: python/pyiceberg/table/__init__.py (hunk 21 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:937]: A `Tuple[str, ...]` would work. I like that

Comment (Fokko) [reply python/pyiceberg/table/__init__.py:937]: I would suggest `Union[str, Tuple[str, ...]]` so you can still pass in a string for adding a field to the root

File: python/pyiceberg/schema.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply python/pyiceberg/schema.py:1115]: I see, sounds good!

File: python/pyiceberg/table/__init__.py (hunk 22 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:961]: That's a nice touch, added

File: python/pyiceberg/schema.py (hunk 17 above)
Comment (hililiwei) [reply python/pyiceberg/schema.py:206]: The old way didn't get the highest field id correctly, so I changed it.

File: python/pyiceberg/table/__init__.py [hunk 25]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        if parent:
+            parent_field = self._schema.find_field(parent, self._case_sensitive)
+            parent_type = parent_field.field_type
+            if isinstance(parent_type, MapType):
+                parent_field = parent_type.value_field
+            elif isinstance(parent_type, ListType):
+                parent_field = parent_type.element_field
+
+            if not parent_field.field_type.is_struct:
+                raise ValueError(f"Cannot add column to non-struct type: {parent}")
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1013]: I think this should also include the parent field name.

File: python/pyiceberg/table/__init__.py [hunk 26]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+            try:
+                exist_field = self._schema.find_field(parent + "." + name, self._case_sensitive)
+            except ValueError:
+                pass
+
+            if exist_field:
+                raise ValueError(f"Cannot add column, name already exists: {parent}.{name}")
+
+            full_name = parent_field.name + "." + name
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1025]: This isn't the full name. The parent field's name is the local name, not its full name.

File: python/pyiceberg/table/__init__.py [hunk 27]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        # assign new IDs in order
+        new_id = self.assign_new_column_id()
+
+        # update tracking for moves
+        self._added_name_to_id[full_name] = new_id
+
+        new_type = assign_fresh_schema_ids(type_var, self.assign_new_column_id)
+        field = NestedField(new_id, name, new_type, not is_optional, doc)
+
+        self._adds.setdefault(parent_id, []).append(field)
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1045]: @Fokko, looks like this doesn't update `_id_to_parent` but the Java implementation does.

Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1045]: Yes, it is currently only appending columns. To keep things smaller, @hililiwei decided to break it up into smaller chunks, which I think is a good idea.

File: python/pyiceberg/table/__init__.py [hunk 28]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+    return Schema(*struct.fields)
+
+
+class _ApplyChanges(SchemaVisitor[IcebergType]):
+    def __init__(self, adds: Dict[int, List[NestedField]]):
+        self.adds = adds
+
+    def schema(self, schema: Schema, struct_result: IcebergType) -> IcebergType:
+        fields = _ApplyChanges.add_fields(schema.as_struct().fields, self.adds.get(TABLE_ROOT_ID))
+        if len(fields) > 0:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1067]: This is not the same as the check in Java. In Java, this checks whether there were any changes to apply. It isn't clear what an empty list is here, but I definitely prefer using None to signal that no changes were necessary rather than an empty list.

File: python/pyiceberg/table/__init__.py [hunk 29]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+            raise ValueError(f"Cannot add fields to map keys: {map_type}")
+
+        value_field: NestedField = map_type.value_field
+        value_type = self.field(value_field, value_result)
+        if value_type is None:
+            raise ValueError(f"Cannot delete value type from map: {value_field}")
+
+        is_value_optional = not map_type.value_required
+
+        if is_value_optional != value_field.required and map_type.value_type == value_type:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1125]: I find this a bit confusing since it negates `value_required` above and then checks `!=` here.

File: python/pyiceberg/table/__init__.py [hunk 30]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        key_id: int = map_type.key_field.field_id
+        if key_id in self.adds:
+            raise ValueError(f"Cannot add fields to map keys: {map_type}")
+
+        value_field: NestedField = map_type.value_field
+        value_type = self.field(value_field, value_result)
+        if value_type is None:
+            raise ValueError(f"Cannot delete value type from map: {value_field}")
+
+        is_value_optional = not map_type.value_required
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1123]: This doesn't come from the right place. In Java, this comes from `updates` where the value may have been made optional.

File: python/pyiceberg/table/__init__.py [hunk 31]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        element_type = self.field(element_field, element_result)
+        if element_type is None:
+            raise ValueError(f"Cannot delete element type from list: {element_field}")
+
+        is_element_optional = not list_type.element_required
+
+        if is_element_optional == element_field.required and list_type.element_type == element_type:
+            return list_type
+
+        return ListType(list_type.element_id, element_type, is_element_optional)
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1111]: @Fokko: This is `is_element_required` so this accidentally flips the boolean. I think we should work with `is_required` throughout this code to avoid needing to negate and accidentally changing whether types are optional.

File: python/pyiceberg/table/__init__.py [hunk 32]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        return field_result
+
+    def list(self, list_type: ListType, element_result: IcebergType) -> IcebergType:
+        element_field: NestedField = list_type.element_field
+        element_type = self.field(element_field, element_result)
+        if element_type is None:
+            raise ValueError(f"Cannot delete element type from list: {element_field}")
+
+        is_element_optional = not list_type.element_required
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1106]: I think this is unnecessary without updates. It can remain as a noop check and placeholder, I guess? In any case, the logic should be fixed below.

File: python/pyiceberg/table/__init__.py [hunk 33]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        if is_element_optional == element_field.required and list_type.element_type == element_type:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1108]: This looks wrong: `is_optional == is_required`.

File: python/pyiceberg/table/__init__.py [hunk 34]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        if has_change:
+            return StructType(*new_fields)
+
+        return struct
+
+    def field(self, field: NestedField, field_result: IcebergType) -> IcebergType:
+        field_id: int = field.field_id
+        if field_id in self.adds:
+            new_fields = self.adds[field_id]
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1092]: Should this use `get` to avoid `KeyError` if the field ID isn't present?

File: python/pyiceberg/table/__init__.py [hunk 35]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+    def primitive(self, primitive: PrimitiveType) -> IcebergType:
+        return primitive
+
+    @staticmethod
+    def add_fields(fields: Tuple[NestedField, ...], adds: Optional[List[NestedField]]) -> List[NestedField]:
+        new_fields: List[NestedField] = []
+        new_fields.extend(fields)
+        if adds:
+            new_fields.extend(adds)
+        return new_fields
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1139]: What about `return tuple(*fields, *adds)`?

File: python/pyiceberg/table/__init__.py [hunk 36]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        if is_value_optional != value_field.required and map_type.value_type == value_type:
+            return map_type
+
+        return MapType(map_type.key_id, map_type.key_field, map_type.value_id, value_type, not is_value_optional)
+
+    def primitive(self, primitive: PrimitiveType) -> IcebergType:
+        return primitive
+
+    @staticmethod
+    def add_fields(fields: Tuple[NestedField, ...], adds: Optional[List[NestedField]]) -> List[NestedField]:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1134]: If adds is None, then I think this should return None.

File: python/pyiceberg/table/__init__.py [hunk 37]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+
+        return struct
+
+    def field(self, field: NestedField, field_result: IcebergType) -> IcebergType:
+        field_id: int = field.field_id
+        if field_id in self.adds:
+            new_fields = self.adds[field_id]
+            if len(new_fields) > 0:
+                fields = _ApplyChanges.add_fields(field_result.fields, new_fields)
+                if len(fields) > 0:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1095]: I think this should be simpler:

```python
new_fields = _ApplyChanges.add_fields(field_result.fields, self.adds.get(field_id))
if new_fields is None:
    return field_result

return StructType(*new_fields)
```

File: python/pyiceberg/table/__init__.py [hunk 38]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+        has_change = False
+        new_fields: List[NestedField] = []
+        for i in range(len(field_results)):
+            type_: Optional[IcebergType] = field_results[i]
+            if type_ is None:
+                has_change = True
+                continue
+
+            field: NestedField = struct.fields[i]
+            new_fields.append(field)
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1082]: Shouldn't this append the type from `field_results`?

File: python/pyiceberg/table/__init__.py [hunk 39]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+    def struct(self, struct: StructType, field_results: List[IcebergType]) -> IcebergType:
+        has_change = False
+        new_fields: List[NestedField] = []
+        for i in range(len(field_results)):
+            type_: Optional[IcebergType] = field_results[i]
+            if type_ is None:
+                has_change = True
+                continue
+
+            field: NestedField = struct.fields[i]
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1081]: @Fokko, we also need to be careful in places like this. This lookup is incorrect because it is matching the field results with the original fields by position. That must be done by field ID.

File: python/pyiceberg/table/__init__.py [hunk 40]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+            new_fields.append(field)
+
+        if has_change:
+            return StructType(*new_fields)
+
+        return struct
+
+    def field(self, field: NestedField, field_result: IcebergType) -> IcebergType:
+        field_id: int = field.field_id
+        if field_id in self.adds:
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:1091]: Should this also check that the field is a struct?

File: python/pyiceberg/table/__init__.py (hunk 29 above)
Comment (hililiwei) [reply python/pyiceberg/table/__init__.py:1125]: Yes, this is only adding columns currently, so this is a bit redundant. It will not be useful until there is an update field. Sorry , it's not clean enough here.

File: python/pyiceberg/table/__init__.py (hunk 30 above)
Comment (hililiwei) [reply python/pyiceberg/table/__init__.py:1123]: Yes, similar to the previous comment.

File: python/pyiceberg/table/__init__.py (hunk 39 above)
Comment (hililiwei) [reply python/pyiceberg/table/__init__.py:1081]: hi @rdblue, I don't understand something. The `field_results` are obtained from the `struct`. And when we get None, the loop will continue. It seems that getting from the position does not cause confusion?

File: python/pyiceberg/table/__init__.py (hunk 40 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1091]: That's a good check, but I think it went okay since `_internal_add_column` only create mappings where the parent is a struct.

File: python/pyiceberg/table/__init__.py (hunk 28 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1067]: I agree, this has been updated in: https://github.com/apache/iceberg/pull/8374/

File: python/pyiceberg/table/__init__.py (hunk 29 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1125]: In https://github.com/apache/iceberg/pull/8374/ I've removed the temporary variable. I that might introduce confusion.

File: python/pyiceberg/table/__init__.py (hunk 31 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1111]: Yes, this is correct, this was already fixed in https://github.com/apache/iceberg/pull/8374/ and I'll make sure to add some more checks

File: python/pyiceberg/table/__init__.py (hunk 36 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1134]: See comment above :)

File: python/pyiceberg/table/__init__.py (hunk 37 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1095]: image

Why no walrus? 

```python
if new_fields := _ApplyChanges.add_fields(field_result.fields, self.adds.get(field_id, [])):
    return field_result

return StructType(*new_fields)
```

File: python/pyiceberg/table/__init__.py [hunk 41]
```diff
@@ -839,3 +887,253 @@ def to_ray(self) -> ray.data.dataset.Dataset:
...
+            return StructType(*fields)
+
+        return struct_result
+
+    def struct(self, struct: StructType, field_results: List[IcebergType]) -> IcebergType:
+        has_change = False
+        new_fields: List[NestedField] = []
+        for i in range(len(field_results)):
+            type_: Optional[IcebergType] = field_results[i]
+            if type_ is None:
```
[the remark this answers is not in the document]
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1077]: Yes, this also confused me a bit

File: python/pyiceberg/table/__init__.py (hunk 38 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:1082]: Yes :)
